core/Enclave/enclave_hsm.cpp

```diff
@@ -458,27 +554,103 @@ sgx_status_t enclave_rsa_encrypt(const uint8_t *cmk_blob,
+    uint32_t max_plaintext_len = ehsm_get_rsa_max_encryption_size(keyspec, padding_mode);
+    if (max_plaintext_len == 0)
+        return SGX_ERROR_INVALID_PARAMETER;
+    if (plaintext_len > max_plaintext_len) {
+        printf("ecall rsa_encrypt plaintext_len is up to %d.\n", max_plaintext_len);
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
 
-    if(rsa == NULL)
-    {
-        printf("failed to load rsa key\n");
+    // check ciphertext and ciphertext_len
+    //
+    if (ciphertext == NULL || ciphertext_len < ehsm_get_rsa_cipher_len(keyspec)) {
+        printf("ecall rsa_encrypt ciphertext len is too small, it should be %d\n", ehsm_get_rsa_cipher_len(keyspec));
         return SGX_ERROR_INVALID_PARAMETER;
     }
+    
+    // check cmk_blob and cmk_blob_size
+    //
+    uint32_t encrypted_rsa_len = ehsm_calc_gcm_data_size(0, ehsm_get_rsa_key_pem_size(keyspec)); // ehsm_get_rsa_key_pem_size() will return INT_MAX if keyspec is invalid
+    
+    if (UINT32_MAX == encrypted_rsa_len) {
+        printf("ecall rsa_encrypt failed to calculate encrypted data size.\n");
+        return SGX_ERROR_UNEXPECTED;
+    }
 
-    switch (keyspec)
-    {
-        case RSA_NO_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_NO_PADDING);
-        case RSA_PKCS1_OAEP_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_PKCS1_OAEP_PADDING);
+    if (cmk_blob == NULL || cmk_blob_size < encrypted_rsa_len) {
+        printf("ecall rsa_encrypt cmk_blob_size is too small.\n");
+        return SGX_ERROR_INVALID_PARAMETER;
     }
 
-    return SGX_SUCCESS;
+    // do rsa public key encrypt
+    //
+    uint8_t* enc_key = NULL;
+    uint8_t* rsa_public_key = NULL;
+    BIO *bio = NULL;
+    RSA* rsa_pubkey = NULL;
+    uint32_t public_key_size;
+    
+    do {
+        // load rsa public key
+        //
+        public_key_size = ehsm_get_rsa_public_key_pem_size(keyspec);
+        
+        enc_key = (uint8_t*)malloc(cmk_blob_size);
+        ret = ehsm_gcm_decrypt(&g_domain_key,
+                              (uint32_t*)&cmk_blob_size, enc_key,
+                              (sgx_aes_gcm_data_ex_t *)cmk_blob);
+        if (ret != SGX_SUCCESS)
+            break;
+
+        rsa_public_key = (uint8_t*)malloc(public_key_size);
+        memcpy(rsa_public_key, enc_key, public_key_size);
+
+        bio = BIO_new_mem_buf(rsa_public_key, -1); // use -1 to auto compute length
+        if (bio == NULL) {
+            printf("failed to load rsa key pem\n");
+            break;
+        }
+
+        rsa_pubkey = PEM_read_bio_RSAPublicKey(bio, NULL, NULL, NULL);
+        if (rsa_pubkey == NULL) {
+            printf("failed to load rsa key\n");
+            break;
+        }
+
+        // make encryption
+        //
+        int encryption_len = RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa_pubkey, padding_mode);
+
+        if (encryption_len == -1 || encryption_len != ciphertext_len)
```

Comment:
see @583, ciphertext_len can be big enough

Reply:
After calling RSA_public_encrypt() correctly, the return value should be RSA_ Size(), which is written in OpenSSL code to check whether it is called correctly.
I'll check this part again